### PR TITLE
[Bugfix] Fix Marlin W4A8 signed scale corruption

### LINKED ISSUE #48905: [Bug]: Marlin W4A8 (int8 activations) corrupts on checkpoints with negative group scales
### Your current environment

<details>
<summary>vLLM v0.25.1, RTX 3090 (Ampere, sm_86)</summary>

Marlin W4A8 (int8-activation) path via `VLLM_MARLIN_INPUT_DTYPE=int8` on a W4A16 checkpoint.
</details>

### 🐛 Describe the bug

With `VLLM_MARLIN_INPUT_DTYPE=int8`, a W4A16 checkpoint whose **weight group scales contain negative values** produces corrupt output (silent — no error, just garbage generations). AutoRound symmetric-int4 checkpoints legitimately emit ~50% negative group scales; these are perfectly valid for W4A16 (fp16 scales), but the int8-activation path corrupts them.

GPTQ / AWQ / llm-compressor recipes emit positive-only scales, which is why this hasn't surfaced.

**Root cause.** `marlin_act_int8_process_scales` (`marlin_utils.py`) packs the per-group scales into int16:

```python
def marlin_act_int8_process_scales(s: torch.Tensor):
    a_scales_scale_factor = 1 / 4096 * s.max().float()
    s = s / s.max() * 4096
    s = s.round().to(torch.int16).view(s.dtype)   # signed int16, negatives preserved
    return s, a_scales_scale_factor
```

so a negative scale is stored as a negative int16 (e.g. `-100 -> 0xFF9C`). But the Marlin kernel's int8-activation scale-application path reads these packed scales as **unsigned** — so `0xFF9C` is interpreted as `65436`, exploding the dequant by ~4 orders of magnitude for every group with a negative scale.

### Reproduce

Self-contained (no checkpoint needed — constructs symmetric int4 weights and toggles the scale sign):

```python
import torch
from vllm import _custom_ops as ops
from vllm.model_executor.layers.quantization.utils.marlin_utils import (
    apply_gptq_marlin_linear, marlin_act_int8_process_scales,
    marlin_make_workspace_new, marlin_permute_scales,
)
from vllm.scalar_type import scalar_types

torch.manual_seed(0); dev = "cuda"; k, n, g, m = 1024, 512, 128, 8
w_q = torch.randint(0, 16, (k, n), dtype=torch.int32, device=dev)
w_pack = torch.zeros(k // 8, n, dtype=torch.int32, device=dev)
for i in range(8): w_pack |= (w_q[i::8] & 0xF) << (4 * i)
empty = torch.empty(0, dtype=torch.int, device=dev); ws = marlin_make_workspace_new(torch.device(dev))
x = torch.randn(m, k, device=dev, dtype=torch.float16) * 0.5

def run(neg_fraction):
    scales = torch.rand(k // g, n, device=dev, dtype=torch.float16) * 0.004 + 0.002
    signs = torch.where(torch.rand_like(scales.float()) < neg_fraction, -1.0, 1.0).half()
    scales = scales * signs
    ref = x.float() @ ((w_q.float() - 8.0) * scales.float().repeat_interleave(g, 0))
    wq = ops.gptq_marlin_repack(w_pack, perm=empty, size_k=k, size_n=n, num_bits=4, is_a_8bit=True)
    s = marlin_permute_scales(scales.clone(), size_k=k, size_n=n, group_size=g, is_a_8bit=True)
    s, igs = marlin_act_int8_process_scales(s)
    out = apply_gptq_marlin_linear(input=x, weight=wq, weight_scale=s, weight_zp=empty, g_idx=empty,
        g_idx_sort_indices=empty, workspace=ws, wtype=scalar_types.uint4b8,
        output_size_per_partition=n, input_size_per_partition=k, is_k_full=True,
        input_global_scale=igs, bias=None, input_dtype=torch.int8)
    rel = ((out.float() - ref).abs() / ref.abs().clamp_min(1e-2)).median().item()
    print(f"{int(neg_fraction*100):3d}% negative group scales -> W4A8 median rel-err {rel:.4f}")

run(0.0)   # positive-only
run(0.5)   # 50% negative
```

Output on v0.25.1 / RTX 3090:

```
  0% negative group scales -> W4A8 median rel-err 0.0078
 50% negative group scales -> W4A8 median rel-err 16.3595
```

### Suggested fix

Read the int16-packed group scales as **signed** in the kernel's int8-activation scale-application path (the values written by `marlin_act_int8_process_scales` are signed int16). Alternatively, `marlin_act_int8_process_scales` could fold the sign into the weight codes (`q -> 16-q` for uint4b8) and store `|scale|`, but the signed read is the direct kernel-side fix.

Impact: any AutoRound-quantized W4A16 checkpoint is silently unusable with the int8-activation toggle.

### PR DESCRIPTION
**Pull request by lazypool**
`marlin_act_int8_process_scales` stores per-group scales as signed `int16_t`, but the kernel's int8-activation scale-application path (`a_type == kS8`) reads them as `uint16_t*`, causing negative scales to be zero-extended to large positive integers. With ~50% negative group scales (common in AutoRound-quantized W4A16 checkpoints) the output is effectively corrupted.

### Fix

Change `reinterpret_cast<uint16_t*>` to `reinterpret_cast<int16_t*>` in two Marlin template headers:

- `csrc/libtorch_stable/quantization/marlin/marlin_template.h`
- `csrc/libtorch_stable/moe/marlin_moe_wna16/marlin_template.h`

### Why this is not duplicating an existing PR

PR #38066 fixes the W4A8-INT compressed_tensors path (int4 weight type not recognized, activations not quantized) but does not address the signed scale read in the kernel template. No other open PR references issue #48905 or the `uint16_t*` to `int16_t*` fix until this PR is created.

### AI Assistance

- AI tooling was used in the implementation and verification of this fix.
- A human has reviewed all changes end-to-end.  The compilation passed on my PC.

Closes #48905

### CODE REVIEW

**Review by lazypool — COMMENTED** [review]
The changes are all reviewed. Could a maintainer or code owner please add the `ready` label to trigger CI? The fix is minimal (8 lines, uint16_t* → int16_t*) and the compilation passed on my PC.

**Comment by gagecane** [comment]
Additional regression boundary for the signed-read fix (AI-assisted analysis; CPU arithmetic checked, CUDA PR not built/tested):

The downstream [syv-ai workaround](https://github.com/syv-ai/qwen38-27b-rtx3090/blob/main/patches/marlin-int8-negative-scales.patch) folds negative scales into signed-int4 codes before packing. That cannot preserve the q=-8 edge: q=-8, scale=-1 gives +8, but sign-flipping and clipping to q=+7 gives +7. Fixing the signed read in this PR avoids needing that lossy transformation.

Compiled and ran a C++17 assertion check with g++ -Wall -Wextra -Werror. For stored int16 values [-32768, -1024, -1, 0, 1, 32767], conversion through uint16 produces [32768, 64512, 65535, 0, 1, 32767]; signed promotion preserves all six values. The clipped-int4 check gives 7 versus the original 8. This only verifies the arithmetic, not device execution.

A useful kernel regression would include mixed-sign group scales, positive controls, and q=-8 with a negative scale, comparing both dense and MoE W4A8 paths to a dequantized reference. The sign-folding workaround must be disabled in that test, or it masks the negative-scale path. The current downstream serving configuration uses W4A16, so its successful chat tests are not evidence for this W4A8 fix.
